features/329-principal-apis/design.md

```diff
@@ -0,0 +1,136 @@
+OAuth. Auth gear should support all identity providers equally well, and
+provide a consistent API for manipulating principals.
+
+
+# Use Cases
+- Get which method is used to log in the current session
+- List all methods to authenticated as a user
+- Add/remove user authentication method
+- Send welcome email to OAuth accounts
+- Populate user metadata according to external profile
+- Create password credentials for users signed up with OAuth
+
+
+# Glossary
+
+**Identity Provider**:
+A method to assert the identity of user. For example, providing a password or
+perform OAuth flow.
+
+**Principal**:
+An identity associated with user. A user can have multiple principals. Since
+the term 'Principal' is defined differently across many frameworks and
+standards, we will instead use a simpler term 'Identity' for public interfaces.
+
+
+# Proposed Design
+
+## Re-authentication for Security
+Many principal manipulation operations are critical for security:
+- Add/remove login ID
+- Link/unlink SSO provider
+- Change password
+
+By default, user are required to re-authenticate before performing these
+operations. This is achieve by checking the issue time of the access token is
+recent.
+
+For change password, we support the common case of providing old password to
+implicitly authenticate the user.
+
+Both the re-auth requirement and time to considered 'recent' can be configured.
+Developer can turn off the re-auth requirement in configuration. However,
+developer should understand the potential security risk.
+
+> Related config: `reauthForSecurity` and `reauthIntervalSeconds`
+> Related API: `changePassword` function
+
+## Current Principal
+We need to provide a way for developer to get the principal is associated with
+the current session:
+- Which login ID is used to login?
+- Which OAuth provider is used to login?
+
+When logging in using a principal, the ID of the principal (i.e. the current
+principal) would be embedded in the access token along with the user ID.
+Information about the current principal would be returned in user object.
+
+> Related API: `identity` field in `User` object
+
+## Principal List
+We need to provide a way for developer to get a list of principals belonged to
+the user:
+- What login IDs the user has?
+- Which OAuth providers had the user linked?
+
+A list of principals belonged to current user would be returned through an API.
+
+> Related API: `listIdentities` function
+
+# Creating Principal
+We need to provide a way for developer to create new principal for the user:
+- Add new login ID
+- Link OAuth account
+
+Each identity provider should have their specific API to create principal. APIs
+that create principals should have common behaviors:
+- Should respect re-authentication configuration
+
+When signup as with multiple login IDs, the first login ID will be used as
+current identity in the newly created session.
+
+> Related API: `addLoginID` function for password identity provider
```

Comment:
link OAuth account?

Reply:
API for SSO is specified in another document, so it is not mentioned here.

Reply:
Similar problem with the above, what is this design document about actually?